src/App.jsx

```diff
@@ -14,33 +16,110 @@
-        Sort by length
-      </button>
+function prepareSortedGoods(goods, sortField, reverse) {
+  const preparedGoods = [...goods];
 
-      <button type="button" className="button is-warning is-light">
-        Reverse
-      </button>
+  if (sortField) {
+    preparedGoods.sort((good1, good2) => {
+      switch (sortField) {
+        case SORT_FIELD_NAME:
+          return good1.localeCompare(good2);
 
-      <button type="button" className="button is-danger is-light">
-        Reset
-      </button>
-    </div>
+        case SORT_FIELD_LENGTH:
+          return good1.length - good2.length;
+
+        default:
+          return 0;
+      }
+    });
+  }
+
+  if (reverse) {
+    preparedGoods.reverse();
+  }
+
+  return preparedGoods;
+}
+
+export const App = () => {
+  const [sortField, setSortField] = useState('');
+  const [firstList, nextList] = useState(goodsFromServer);
+
+  const [reverse, setReverse] = useState(false);
+  const visibleGoods = prepareSortedGoods(firstList, sortField, reverse);
 
-    <ul>
-      <li data-cy="Good">Dumplings</li>
-      <li data-cy="Good">Carrot</li>
-      <li data-cy="Good">Eggs</li>
-      <li data-cy="Good">Ice cream</li>
-      <li data-cy="Good">Apple</li>
-      <li data-cy="Good">...</li>
-    </ul>
-  </div>
-);
+  return (
+    <div className="section content">
+      <div className="buttons">
+        <button
+          type="button"
+          onClick={() => {
+            nextList(ENTER_GOODS);
+            setSortField(SORT_FIELD_NAME);
+          }}
+          className={cn('button is-info', {
+            'is-light': sortField !== SORT_FIELD_NAME,
+          })}
+        >
+          Sort alphabetically
+        </button>
+
+        <button
+          type="button"
+          onClick={() => {
+            nextList(ENTER_GOODS);
+            setSortField(SORT_FIELD_LENGTH);
+          }}
+          className={cn('button is-success', {
+            'is-light': sortField !== SORT_FIELD_LENGTH,
+          })}
+        >
+          Sort by length
+        </button>
+
+        <button
+          type="button"
+          onClick={() => {
+            setSortField(sortField);
+            // eslint-disable-next-line no-unused-expressions
+            !reverse ? setReverse(true) : setReverse(false);
+          }}
+          className={cn('button is-warning', {
+            'is-light': reverse === false,
+          })}
+        >
+          Reverse
+        </button>
+
+        {!visibleGoods.every((good, index) => good === ENTER_GOODS[index]) && (
+          <button
+            type="button"
+            onClick={() => {
+              setSortField('');
+              nextList(ENTER_GOODS);
+              setReverse(false);
+            }}
+            className={cn('button is-danger is-light')}
+          >
+            Reset
+          </button>
+        )}
+      </div>
+
+      <div>
+        <ul>
+          {visibleGoods.map((good, index) => (
```

Comment:
Each element in the list should have a unique 'key' prop to help React identify which items have changed, are added, or are removed. Using the index as a key is not recommended if the list can change. Consider using a unique identifier for each item if possible.